src/compiler/checker.ts

```diff
@@ -22763,6 +22763,10 @@ namespace ts {
                         getUnionType([extractDefinitelyFalsyTypes(strictNullChecks ? leftType : getBaseTypeOfLiteralType(rightType)), rightType]) :
                         leftType;
                 case SyntaxKind.BarBarToken:
+                    leftType = leftType.flags & TypeFlags.IndexedAccess ?
+                        any(getPropertiesOfType((<IndexedAccessType>leftType).objectType), prop => !!(prop.flags & SymbolFlags.Optional)) ?

```

Comment:
I don't think this is right - checking is `any` prop in the indexed access's type is optional and then returning `never` for its type if so seems very wrong - the potential type of the LHS is completely lost here, and it loses the nuance of the individual property accessed. This needs #29317 to be correctly solved - the LHS type should be `leftType & not 0 & not null & not undefined & not ""`, which handles generics appropriately.

Reply:
The fact that we don't have more tests this change, as-is, breaks is actually a little disturbing.